### ISSUE TITLE
Exception during migration creates some tables

### ISSUE THREAD

**Issue by pate** [yogthos/migratus#74]
I had a syntax error in my migration spanning multiple statements. When I ran (migratus/up config 2016...), Migratus threw an exception. After fixing the syntax error and re-ran `migratus/up`, it turns out half the tables were created, but nothing was written to the migrations table, leaving my DB in a dreadful state on the river styx.

Surely Migratus should rollback all previous statements in the event of an exception? Or am I doing something wrong?

Here was the exception:

```
clojure.lang.Compiler$CompilerException: org.postgresql.util.PSQLException: ERROR: relation "users" already exists, compiling:(/~/src/my-project/core/migrations.clj:33:3)
      org.postgresql.util.PSQLException: ERROR: relation "users" already exists
```

**Comment by yogthos** [comment]
You have to explicitly separate multiple statements using the `--;;` notation as described [here](https://github.com/yogthos/migratus#multiple-statements).

**Comment by pate** [comment]
I am using the separator as specified in the docs, e.g.:

```
create table users (
  id serial primary key not null,
  created_on timestamp not null default now(),
  full_name text not null,
  email text not null,
  is_active boolean not null default true
)
--;;
create table identities (
  id serial primary key not null,
  created_on timestamp not null default now(),
  user_id int references users(id) not null,
  provider text not null,
  username text not null,
  password_hash text null,
  password_salt text null,
  is_active boolean not null default true
)
--;;
create table sessions (
  id serial primary key not null,
  user_id int references users(id) not null,
  created_on timestamp not null default now(),
  expires_on timestamp not null,
  is_active boolean not null default true
)
```

**Comment by yogthos** [comment]
What version of the library are you using. There was a fix to rethrow the exception in this commit https://github.com/yogthos/migratus/commit/29c4753bf8a454b0d31b0f379a8a63b4c9cdd618 and running up/down should be transaction. So, if you're on a version before `0.8.15` you'd see this issue.

**Comment by pate** [comment]
I'm on `0.8.16`.

**Comment by yogthos** [comment]
Hmm another question is what db you're running against.

**Comment by adambaker** [comment]
I'm seeing this problem too, using migratus `0.8.25` with postgresql version `9.4.8`, jdbc driver version `9.4-1206-jdbc4`. Here's a pretty minimal script that fails:

```
create table users (
  id serial primary key
);
--;;
create table adults (
  id integer references users primary key,
  email text not null unique
);
--;;
create table students (
create table students (
  id integer references users primary key,
  email text unique,
  grade smallint not null
);
```

Running this generates the expected sql syntax error, but rather than rollback a transaction it leaves the `adults` and `users` tables in the database.

**Comment by pate** [comment]
@yogthos, I'm running Postgres `9.5.0.0`.

**Comment by pate** [comment]
From skimming the code, it seems that the only code that runs in a transaction, is each individual up/down statement:
https://github.com/yogthos/migratus/blob/7aedc8b57ecd79cef076e46b6c5550442cf510fb/src/migratus/database.clj#L68

You probably want to wrap the entire sequence of migrations in a transaction, like here:
https://github.com/yogthos/migratus/blob/7aedc8b57ecd79cef076e46b6c5550442cf510fb/src/migratus/core.clj#L53

**Comment by yogthos** [comment]
The core is data store agnostic, so it wouldn't be the place to wrap JDBC specific transactions. However, each transaction is meant to be independent in any case. It might make sense to stop and as soon as an up statement returns false though, since at that point it wouldn't make sense to apply the rest of transactions.

**Comment by yogthos** [comment]
[This](https://github.com/yogthos/migratus/blob/77cfb6f9d3437b835bb5954b3f53a82f3a410bcc/src/migratus/core.clj#L53) would be my proposed solution here. The migrations will stop being applied as soon as any migration fails.

**Comment by pate** [comment]
Ah yes, multiple statements in a migration should run in a transaction - not all migrations (as I badly suggested earlier). And if any statement fails, the transaction for that migration should be rolled back.

**Comment by yogthos** [comment]
So, that should be the current behavior. I do think it makes sense to stop further when  migration does fail. That's the purpose of the new update.

**Comment by samroberton** [comment]
We've also seen this behaviour:  a migration script where the second statement fails (causing an exception which propagates back out of `migratus` to our calling code) somehow commits the effect of the first statement to the database (but with no entry in `schema_migrations` for the partially-applied migration).

We're running `migratus 0.8.20`, on Postgres 9.5, with version 9.4.1208 of the Postgres JDBC driver, and `clojure.java.jdbc 0.5.8`.

**Comment by samroberton** [comment]
Found the error:  it was incorrect use of `sql/with-db-transaction` (opening the transaction, but then proceeding to use the db-spec itself as the DB connection parameter, instead of the transaction), which was fixed (for up migrations at least) by this commit:  f24349a.  Ie for up migrations, at least, this is fixed in 0.8.28.

Down migrations still incorrectly use `db` instead of `t-con` as the DB connection parameter.  PR incoming.

**Comment by yogthos** [comment]
Thanks for the pr, just pushed out a new version with the fix.
